### PR TITLE
Make LinkMonitor CLA compatible

### PR DESCRIPTION
**Pull request by fbac**
- `addToWatchList` function has to be CL Automation compatible.
- `removeFromWatchList` function has to properly clear the address from the watchList and set it as not active.
- Contract now has admin and executor roles, so the admin can give permissions to different CLA Upkeeps.
- Roles can be revoked.

### CODE REVIEW

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 1]
```diff
@@ -2,10 +2,9 @@
 
 pragma solidity 0.8.19;
 
+import "@openzeppelin/contracts/access/AccessControlEnumerable.sol";
```
Comment (RensR) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:5]: Please follow the [style guide](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/STYLE_GUIDE.md)

All imports should be explicit and they should follow a set format

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 2]
```diff
@@ -70,23 +72,42 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
...
     LINK_TOKEN = IERC20(linkTokenAddress);
     setMinWaitPeriodSeconds(minWaitPeriodSeconds);
     setMaxPerform(maxPerform);
     setMaxCheck(maxCheck);
     setUpkeepInterval(upkeepInterval);
   }
 
+  /// @notice Grants an address an executor role
+  /// @param executor address to grant executor role to
+  function granExecutorRole(address executor) public onlyRole(ADMIN_ROLE) {
```
Comment (RensR) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:99]: typo, missing t

**Review by RensR — COMMENTED** [review]
Missing a lot of test coverage

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 3]
```diff
@@ -118,6 +139,50 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
...
+    // there's no need to remove any other address that's not an onRamp
+    if (dstChainSelector > 0 && bytes(abi.encodePacked(oldAddress)).length > 0) {
+      removeFromWatchList(oldAddress);
+    }
+    // only add the new address if it's not 0x0
+    if (targetAddress != address(0)) {
+      s_onRampAddresses[dstChainSelector] = targetAddress;
+      s_targets[targetAddress] = MonitoredAddress({
+        isActive: true,
+        minBalance: 1,
```
Comment (RensR) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:161]: All magic values should be constants (note 0 is not magic, 1 and 9 are)

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 4]
```diff
@@ -33,7 +32,10 @@ interface ILinkAvailable {
 ///  this is a "trusless" upkeep, meaning it does not trust the caller of performUpkeep;
 /// we could save a fair amount of gas and re-write this upkeep for use with Automation v2.0+,
 /// which has significantly different trust assumptions
-contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationCompatibleInterface {
+contract LinkAvailableBalanceMonitor is AccessControlEnumerable, AutomationCompatibleInterface {
+  bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
```
[the remark this answers is not in the document]
Comment (fbac) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:36]: Yes, we can use DEFAULT_ADMIN_ROLE with no problem and no change of this contract's behavior. Since we are also creating EXECUTOR_ROLE I just went ahead defining explicitly the two roles used on this contract, just because it's more intuitive.
I don't have a problem in just using the default one if you believe it's more readable/logical.

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol (hunk 1 above)
Comment (fbac) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol]: Done!

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 5]
```diff
@@ -118,6 +141,50 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
...
+      s_targets[targetAddress] = MonitoredAddress({
+        isActive: true,
+        minBalance: DEFAULT_MIN_BALANCE,
+        topUpAmount: DEFAULT_TOP_UP_AMOUNT,
+        lastTopUpTimestamp: 0
+      });
+      s_watchList.push(targetAddress);
+    } else {
+      // if the address is 0x0, it means the onRamp has ben decomissioned and has to be cleaned
+      delete s_onRampAddresses[dstChainSelector];
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:170]: this feels like a non-obvious use case for the addToWatchList() function. Would it make more sense to move this functionality to another public function like `decomissionOnRamp(address target)` or something similar? Then in `addToWatchlist(..)` you could `require(targetAddess != address(0))`.

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 6]
```diff
@@ -2,10 +2,9 @@
 
 pragma solidity 0.8.19;
 
+import {AccessControl} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/access/AccessControl.sol";
```
Comment (matYang) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:5]: 3rd party imports should come last, pls refer to the latest Solidity style guide on imports

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 7]
```diff
@@ -63,30 +62,54 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
...
     LINK_TOKEN = IERC20(linkTokenAddress);
     setMinWaitPeriodSeconds(minWaitPeriodSeconds);
     setMaxPerform(maxPerform);
     setMaxCheck(maxCheck);
     setUpkeepInterval(upkeepInterval);
   }
 
+  /// @notice Grants an address an executor role
+  /// @param executor address to grant executor role to
+  function granExecutorRole(address executor) public onlyRole(ADMIN_ROLE) {
```
Comment (matYang) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:101]: similar to below, OZ has a grantRole public function

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 8]
```diff
@@ -33,7 +32,7 @@ interface ILinkAvailable {
 ///  this is a "trusless" upkeep, meaning it does not trust the caller of performUpkeep;
 /// we could save a fair amount of gas and re-write this upkeep for use with Automation v2.0+,
 /// which has significantly different trust assumptions
-contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationCompatibleInterface {
```
Comment (matYang) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:36]: can you commend on security implication if Pausable is removed

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 9]
```diff
@@ -63,30 +62,54 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
...
     setMaxPerform(maxPerform);
     setMaxCheck(maxCheck);
     setUpkeepInterval(upkeepInterval);
   }
 
+  /// @notice Grants an address an executor role
+  /// @param executor address to grant executor role to
+  function granExecutorRole(address executor) public onlyRole(ADMIN_ROLE) {
+    if (executor == address(0)) revert InvalidAddress(executor);
+    _setupRole(EXECUTOR_ROLE, executor);
```
[the remark this answers is not in the document]
Comment (fbac) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:103]: Updated to _grantRole.

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 10]
```diff
@@ -118,6 +141,50 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
...
+  /// @param dstChainSelector carries a non-zero value in case the targetAddress is an onRamp, otherwise it carries a 0
+  /// @dev this function has to be compatible with the event onRampSet(address, dstChainSelector) emitted by
+  /// the CCIP router. Important detail to know is this event is also emitted when an onRamp is decomissioned,
+  /// in which case it will carry the proper dstChainSelector along with the 0x0 address
+  function addToWatchList(address targetAddress, uint64 dstChainSelector) public onlyRoleOrAdminRole(EXECUTOR_ROLE) {
+    if (s_targets[targetAddress].isActive) revert DuplicateAddress(targetAddress);
+    address oldAddress = s_onRampAddresses[dstChainSelector];
+    // if targetAddress is an existing onRamp, there's a need of cleaning the previous onRamp associated to this dstChainSelector
+    // there's no need to remove any other address that's not an onRamp
+    if (dstChainSelector > 0 && bytes(abi.encodePacked(oldAddress)).length > 0) {
```
[the remark this answers is not in the document]
Comment (fbac) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:155]: Not exactly I think, on this line I want to make sure the dstChainSelector is positive and the oldAddress contains bytes (it exists in the mapping s_onRampAddresses)
If the address does not exist on s_onRampAddresses I assume oldAddress would be 0 bytes, not address(0).

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol (hunk 5 above)
Comment (fbac) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:170]: The problem behind this, which I agree with you that this is totally non-obvious, is that this function has to be triggered by a CL Automation upkeep, and while doing so preserving the sanity of the watchlist.
CL Automation would trigger in onRampSet(dstChainSelector, address) events, and the gotcha here is when an onRamp is decomissioned the onRampSet carries the dstChainSelector and address(0), meaning we have to remove the onRamp completely. This is the only corner case this would be triggered.

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol (hunk 6 above)
Comment (fbac) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:5]: Addressed!

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol (hunk 8 above)
Comment (fbac) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:36]: I removed Pausable due to a problem with solidity, as it's solved now I'm reintroducing it again.

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol (hunk 7 above)
Comment (fbac) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:101]: Agree, removing them.

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 11]
```diff
@@ -63,30 +62,54 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
...
+    _setupRole(EXECUTOR_ROLE, executor);
+  }
+
+  /// @notice Revokes the executor role from an address
+  /// @param executor address to revoke executor role from
+  function revokeExecutorRole(address executor) public onlyRole(ADMIN_ROLE) {
+    if (executor == address(0)) revert InvalidAddress(executor);
+    _revokeRole(EXECUTOR_ROLE, executor);
+  }
+
```
[the remark this answers is not in the document]
Comment (fbac) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:112]: I agree, removing them.

**Comment by fbac** [comment]
> I don't think the test changes quite match the functionality changes. Ex I don't see anything in the tests about the executor role? Also some new edge cases were added to `addToWatchlist()` that should have test cases.
> 
> Also, this contract is very CCIP specific. I think you might want to either (A) make a copy of the original `LinkAvailableBalanceMonitor` contract or (B) rename this to something like `CCIPOnRampBalanceMonitor` or similar. I think preference for (A) as iirc this contract was _supposed_ to serve any contract that implements the `linkAvailableForPayment()` function (ex VRF) but these changes only make sense for CCIP.

Working on the unit tests, I'll add them asap.

Regarding the CCIP specific part, I agree. The idea here is this contract will watch over every kind of contract (vrf, data feed proxies..) but also over CCIP onRamps, and in this last case it will be triggered automatically through the onRampSet event (using Automation). So that's why some of the logic has to be changed while it's still capable of dealing with any other contract type.
Another solution is to create a middleman contract that manages the s_onRamp list and triggers an addition on this contract's watchlist if everything is ok. It would increase the attack surface though. 
What do you think?

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (RyanRHall) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:143]: unset values in solidity always default to their "zero value" - for for address types that is `address(0)`. So even if the underlying storage has never been set for a particular address storage variable, it will _still_ be equal to `address(0)`.

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol (hunk 5 above)
Comment (RyanRHall) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:158]: ah I see! Maybe just a different function name would suffice then? Something that more accurately describes the intended use case.

**Comment by RyanRHall** [comment]
> Another solution is to create a middleman contract that manages the s_onRamp list and triggers an addition on this contract's watchlist if everything is ok. It would increase the attack surface though.
What do you think?

Personally, I would learn towards having multiple, use-case-specific contracts in the codebase (especially as upkeeps) than a single generic contract. I think that feels safer to me. I would just go all out and make this the `CCIPOnRampUpkeep` or whatnot.

File: contracts/scripts/native_solc_compile_all_automation [hunk 12]
```diff
@@ -77,3 +77,11 @@ compileContract tests/VerifiableLoadStreamsLookupUpkeep.sol
...
 compileContract tests/AutomationConsumerBenchmark.sol
 compileContract tests/StreamsLookupUpkeep.sol
+
+SOLC_VERSION="0.8.19"
+
+solc-select install $SOLC_VERSION
+solc-select use $SOLC_VERSION
+export SOLC_VERSION=$SOLC_VERSION
+
+compileContract automation/upkeeps/LinkAvailableBalanceMonitor.sol
```
Comment (RyanRHall) [inline contracts/scripts/native_solc_compile_all_automation:87]: why is this necessary? We usually only make go wrappers for contracts that the core node needs to interact with.

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 13]
```diff
@@ -33,7 +34,7 @@ interface ILinkAvailable {
 ///  this is a "trusless" upkeep, meaning it does not trust the caller of performUpkeep;
 /// we could save a fair amount of gas and re-write this upkeep for use with Automation v2.0+,
 /// which has significantly different trust assumptions
-contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationCompatibleInterface {
+contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AccessControl, AutomationCompatibleInterface {
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:37]: Do we need both `ConfirmedOwner` _and_ `AccessControl`? I think the latter is a "super set" of the former.

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol [hunk 14]
```diff
@@ -63,23 +64,33 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
     bool isActive;
   }
 
+  bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+  bytes32 public constant EXECUTOR_ROLE = keccak256("EXECUTOR_ROLE");
+  uint96 private constant DEFAULT_TOP_UP_AMOUNT = 9000000000000000000;
+  uint96 private constant DEFAULT_MIN_BALANCE = 1000000000000000000;
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:70]: can we call these `DEFAULT_TOP_UP_AMOUNT_JULES` and `DEFAULT_MIN_BALANCE_JULES`?

Comment (RyanRHall) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:70]: ("jules" is like "wei" for LINK)

**Review by RyanRHall — CHANGES_REQUESTED** [review]

File: contracts/scripts/native_solc_compile_all_automation (hunk 12 above)
Comment (fbac) [reply contracts/scripts/native_solc_compile_all_automation:87]: it's there because I used this script to compile the contract and added it by mistake. Removing!

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol (hunk 13 above)
Comment (fbac) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:37]: I thought the same, but after some testing it seems ConfirmedOwner is needed by Pausable. If the contract doesn't implement ConfirmedOwner it would fail. I'll be testing this a bit more though.

File: contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (fbac) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:143]: Gotcha! Thanks for educating. :)

[the remark this answers is not in the document]
Comment (RyanRHall) [reply contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:143]: Oh I think we should still make this change! I think `oldAddress != address(0)` is cleaner

**Review by RyanRHall — CHANGES_REQUESTED** [review]
